File: app/src/main/java/cn/com/woong/readhub/ui/news/NewsFragment.java
package cn.com.woong.readhub.ui.news;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.PagerAdapter;
import android.support.v4.view.ViewPager;
import android.text.TextUtils;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

import butterknife.BindView;
import cn.com.woong.readhub.R;
import cn.com.woong.readhub.base.BaseFragment;
import cn.com.woong.readhub.bean.NewsMo;
import cn.com.woong.readhub.ui.widget.newsview.NewsView;
import cn.com.woong.readhub.utils.CommonUtils;
import dagger.android.support.AndroidSupportInjection;

/**
 * Created by wong on 2018/3/9.
 */

public class NewsFragment extends BaseFragment<NewsPresenter> implements NewsContract.View {
    @BindView(R.id.news_tab_layout)
    TabLayout newsTabLayout;
    @BindView(R.id.news_view_pager)
    ViewPager newsViewPager;

    private NewsViewPagerAdapter mNewsViewPagerAdapter;
    private ArrayList<NewsView> mNewsViewList = new ArrayList<>();
    private ArrayList<String> mPublishDateList = new ArrayList<>();
    private String[] mTabTitle = {"科技动态", "开发者资讯", "区块链快讯"};
    private int mCurrentTabPos = 0;

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_news;
    }

    @Override
    protected void initView(View view) {
        initViewPager();
    }

    @Override
    protected void initData() {
        if (mPresenter != null) {
            mPresenter.getTechNews("");
        }

        newsTabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                mCurrentTabPos = tab.getPosition();
                if (mPresenter != null) {
                    switch (tab.getPosition()) {
                        case 0:
                            mPresenter.getTechNews("");
                            break;
                        case 1:
                            mPresenter.getDevelopNews("");
                            break;
                        case 2:
                            mPresenter.getBlockchainNews("");
                            break;
                        default:
                            break;
                    }
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {}

            @Override
            public void onTabReselected(TabLayout.Tab tab) {}
        });
    }

    private void initViewPager() {
        for (int i = 0, count = 3; i < count; i++) {
            final int pos = i;
            NewsView newsView = new NewsView(getActivity());
            newsView.setOnNewsListener(new NewsView.OnNewsListener() {
                @Override
                public void onNewsLoadMore() {
                    if (mPresenter != null) {
                        switch (pos) {
                            case 0:
                                mPresenter.getTechNews(mPublishDateList.get(0));
                                break;
                            case 1:
                                mPresenter.getDevelopNews(mPublishDateList.get(1));
                                break;
                            case 2:
                                mPresenter.getBlockchainNews(mPublishDateList.get(2));
                                break;
                            default:
                                break;
                        }
                    }
                }

                @Override
                public void onNewsRefreshing() {
                    if (mPresenter != null) {
                        mPublishDateList.set(pos, "");
                        switch (pos) {
                            case 0:
                                mPresenter.getTechNews("");
                                break;
                            case 1:
                                mPresenter.getDevelopNews("");
                                break;
                            case 2:
                                mPresenter.getBlockchainNews("");
                                break;
                            default:
                                break;
                        }
                    }
                }
            });
            mNewsViewList.add(newsView);
            mPublishDateList.add("");
        }

        mNewsViewPagerAdapter = new NewsViewPagerAdapter();
        newsViewPager.setAdapter(mNewsViewPagerAdapter);
        newsTabLayout.setupWithViewPager(newsViewPager);
    }

    public void refreshNewsData() {
        if (mPresenter != null) {
            mPublishDateList.set(mCurrentTabPos, "");
            switch (mCurrentTabPos) {
                case 0:
                    mPresenter.getTechNews("");
                    break;
                case 1:
                    mPresenter.getDevelopNews("");
                    break;
                case 2:
                    mPresenter.getBlockchainNews("");
                    break;
                default:
                    break;
            }
        }
    }

    @Override
    public void updateTechNews(String publishDate, ArrayList<NewsMo> newsMos) {
        mPublishDateList.set(0, String.valueOf(CommonUtils.getTimeStampByReadhubDateString(
                newsMos.get(newsMos.size() - 1).publishDate)));
        mNewsViewList.get(0).updateNews(TextUtils.isEmpty(publishDate), newsMos);
    }

    @Override
    public void updateDevelopNews(String publishDate, ArrayList<NewsMo> newsMos) {
        mPublishDateList.set(1, String.valueOf(CommonUtils.getTimeStampByReadhubDateString(
                newsMos.get(newsMos.size() - 1).publishDate)));
        mNewsViewList.get(1).updateNews(TextUtils.isEmpty(publishDate), newsMos);
    }

    @Override
    public void updateBlockchainNews(String publishDate, ArrayList<NewsMo> newsMos) {
        mPublishDateList.set(2, String.valueOf(CommonUtils.getTimeStampByReadhubDateString(
                newsMos.get(newsMos.size() - 1).publishDate)));
        mNewsViewList.get(2).updateNews(TextUtils.isEmpty(publishDate), newsMos);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mNewsViewList.clear();
        mPublishDateList.clear();
    }

    private class NewsViewPagerAdapter extends PagerAdapter {
        @Override
        public int getCount() {
            return mNewsViewList.size();
        }

        @NonNull
        @Override
        public Object instantiateItem(@NonNull ViewGroup container, int position) {
            container.addView(mNewsViewList.get(position));
            return mNewsViewList.get(position);
        }

        @Override
        public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
            container.removeView(mNewsViewList.get(position));
        }

        @Override
        public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
            return view == object;
        }

        @Nullable
        @Override
        public CharSequence getPageTitle(int position) {
            return mTabTitle[position];
        }
    }
}

File: app/src/main/java/cn/com/woong/readhub/ui/news/NewsFragment.kt
package cn.com.woong.readhub.ui.news

import android.content.Context
import android.view.ViewGroup
import android.support.v4.view.PagerAdapter
import android.text.TextUtils
import cn.com.woong.readhub.utils.CommonUtils
import cn.com.woong.readhub.bean.NewsMo
import cn.com.woong.readhub.ui.widget.newsview.NewsView
import android.support.design.widget.TabLayout
import android.view.View
import cn.com.woong.readhub.R
import cn.com.woong.readhub.base.BaseFragment
import kotlinx.android.synthetic.main.fragment_news.*

class NewsFragment : BaseFragment<NewsPresenter>(), NewsContract.View {
    private var mNewsViewPagerAdapter: NewsViewPagerAdapter? = null
    private val mNewsViewList = arrayListOf<NewsView>()
    private val mPublishDateList = arrayListOf<String>()
    private val mTabTitle = arrayOf("科技动态", "开发者资讯", "区块链快讯")
    private var mCurrentTabPos = 0

    override fun getLayoutId(): Int {
        mPresenter = NewsPresenter()
        return R.layout.fragment_news
    }

    override fun initView(view: View?) {
        initViewPager()
    }

    override fun initData() {
        mPresenter?.getTechNews("")

        news_tab_layout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                mCurrentTabPos = tab.position

                mPresenter?.let {
                    when (tab.position) {
                        0 -> mPresenter.getTechNews("")
                        1 -> mPresenter.getDevelopNews("")
                        2 -> mPresenter.getBlockchainNews("")
                        else -> { }
                    }
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}

            override fun onTabReselected(tab: TabLayout.Tab) {}
        })
    }

    private fun initViewPager() {
        var i = 0
        val count = 3
        while (i < count) {
            val pos = i
            val newsView = NewsView(activity as Context)
            newsView.setOnNewsListener(object : NewsView.OnNewsListener {
                override fun onNewsLoadMore() {
                    mPresenter?.let {
                        when (pos) {
                            0 -> mPresenter.getTechNews(mPublishDateList.get(0))
                            1 -> mPresenter.getDevelopNews(mPublishDateList.get(1))
                            2 -> mPresenter.getBlockchainNews(mPublishDateList.get(2))
                            else -> {
                            }
                        }
                    }
                }

                override fun onNewsRefreshing() {
                    mPresenter?.let {
                        mPublishDateList.set(pos, "")
                        when (pos) {
                            0 -> mPresenter.getTechNews("")
                            1 -> mPresenter.getDevelopNews("")
                            2 -> mPresenter.getBlockchainNews("")
                            else -> {
                            }
                        }
                    }
                }
            })
            mNewsViewList.add(newsView)
            mPublishDateList.add("")
            i++
        }

        mNewsViewPagerAdapter = NewsViewPagerAdapter()
        news_view_pager.adapter = mNewsViewPagerAdapter
        news_tab_layout.setupWithViewPager(news_view_pager)
    }

    override fun onResume() {
        super.onResume()
        for (newsView in mNewsViewList) {
            newsView.loadMoreComplete()
        }
    }

    fun refreshNewsData() {
        if (mPresenter != null) {
            mPublishDateList.set(mCurrentTabPos, "")
            mNewsViewList.get(mCurrentTabPos).scrollToPosition(0)
            when (mCurrentTabPos) {
                0 -> mPresenter.getTechNews("")
                1 -> mPresenter.getDevelopNews("")
                2 -> mPresenter.getBlockchainNews("")
                else -> {
                }
            }
        }
    }

    override fun updateTechNews(publishDate: String, newsMos: ArrayList<NewsMo>) {
        mPublishDateList.set(0, CommonUtils.getTimeStampByReadhubDateString(
                newsMos[newsMos.size - 1].publishDate).toString())
        mNewsViewList.get(0).updateNews(TextUtils.isEmpty(publishDate), newsMos)
    }

    override fun updateDevelopNews(publishDate: String, newsMos: ArrayList<NewsMo>) {
        mPublishDateList.set(1, CommonUtils.getTimeStampByReadhubDateString(
                newsMos[newsMos.size - 1].publishDate).toString())
        mNewsViewList.get(1).updateNews(TextUtils.isEmpty(publishDate), newsMos)
    }

    override fun updateBlockchainNews(publishDate: String, newsMos: ArrayList<NewsMo>) {
        mPublishDateList.set(2, CommonUtils.getTimeStampByReadhubDateString(
                newsMos[newsMos.size - 1].publishDate).toString())
        mNewsViewList.get(2).updateNews(TextUtils.isEmpty(publishDate), newsMos)
    }

    override fun onDestroy() {
        super.onDestroy()
        mNewsViewList.clear()
        mPublishDateList.clear()
    }

    private inner class NewsViewPagerAdapter : PagerAdapter() {
        override fun getCount(): Int {
            return mNewsViewList.size
        }

        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            container.addView(mNewsViewList.get(position))
            return mNewsViewList.get(position)
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            container.removeView(mNewsViewList.get(position))
        }

        override fun isViewFromObject(view: View, `object`: Any): Boolean {
            return view === `object`
        }

        override fun getPageTitle(position: Int): CharSequence? {
            return mTabTitle[position]
        }
    }
}
